ONSdigital/dp-kafka
# Fixed state channels data race

### What

A new data-race has been identified by running `dp-cantabular-metadat-exporter` component tests, where a chan wait operation (read) was concurrently trying to access a channel that might have been written by another go-routine (state transition).

Trello card: https://trello.com/c/3heYypYs/5470-bugfix-dp-kafka-data-race

This PR fixes the race condition by introducing a new struct, wrapping state channels and a corresponding mutex:

- State channels fix:
  - created struct containing the state channel and a mutex
  - Enter: closes the channel (with read-only mutex lock)
  - Leave: resets the channel (with a full mutex lock)
  - Wait: waits for the channel to be closed (with a read-only lock)

Why I think a deadlock is not possible in StateChannels (proved by new unit-test):
- If we are in state A and wait for state A, the call will exist straight away
- If we are in state A and wait for state B, the call will block (read-only) on lock B 
- When a state transition happens from A->B, then lock A will be acquired (full) and B will be acquired(read-only)

Hence, when the transition from A to B happens:
- we will never have any lock on A, because any call to Wait will return straight away
- we might have read-only locks on B, but the transition to B will only acquire a read-only lock because it does not require resetting any channel, so it will be allowed to execute, and in turn will release any other go-routine that was waiting on B.

Tested against `dp-cantabular-metadat-exporter` [component tests with `-race` flag set](https://github.com/ONSdigital/dp-cantabular-metadata-exporter/pull/32)

### How to review

- Try to challenge the solution (can you find any corner case scenario where we create a dead-lock or a data race?)
- Make sure code changes make sense
- Make sure unit tests pass

### Who can review

anyone (@redhug1  probably)
